runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -0,0 +1,146 @@
+  public static final String DROPPED_DUE_TO_LATENESS_COUNTER = "StatefulParDoDropped";
+
+  private DoFn<InputT, OutputT> fn;
+  private final SimpleDoFnRunner<InputT, OutputT> doFnRunner;
+  private final StepContext stepContext;
+  private final WindowingStrategy<?, ?> windowingStrategy;
+  private final Aggregator<Long, Long> droppedDueToLateness;
+  private final Coder<BoundedWindow> windowCoder;
+  private final DoFnSignature signature;
+
+  @SuppressWarnings("unchecked")
+  public StatefulDoFnRunner(
+      DoFn<InputT, OutputT> fn,
+      SimpleDoFnRunner<InputT, OutputT> doFnRunner,
+      StepContext stepContext,
+      AggregatorFactory aggregatorFactory,
+      WindowingStrategy<?, ?> windowingStrategy) {
+    this.fn = fn;
+    this.doFnRunner = doFnRunner;
+    this.stepContext = stepContext;
+    this.windowingStrategy = windowingStrategy;
+    this.signature = DoFnSignatures.getSignature(fn.getClass());
+    windowCoder = (Coder<BoundedWindow>) windowingStrategy.getWindowFn().windowCoder();
+    this.droppedDueToLateness = aggregatorFactory.createAggregatorForDoFn(
+        fn.getClass(), stepContext, DROPPED_DUE_TO_LATENESS_COUNTER,
+        Sum.ofLongs());
+  }
+
+  @Override
+  public void startBundle() {
+    doFnRunner.startBundle();
+  }
+
+  @Override
+  public void processElement(WindowedValue<InputT> compressedElem) {
+
+    TimerInternals timerInternals = stepContext.timerInternals();
+
+    // StatefulDoFnRunner is always observes window
+    for (WindowedValue<InputT> value : compressedElem.explodeWindows()) {
+
+      BoundedWindow window = value.getWindows().iterator().next();
+      Instant gcTime = window.maxTimestamp().plus(windowingStrategy.getAllowedLateness());
+
+      if (!dropLateData(window)) {
+        timerInternals.setTimer(StateNamespaces.window(windowCoder, window),
+            GC_TIMER_ID, gcTime, TimeDomain.EVENT_TIME);
+        doFnRunner.invokeProcessElement(value);
+      }
+    }
+  }
+
+  private boolean dropLateData(BoundedWindow window) {
+    TimerInternals timerInternals = stepContext.timerInternals();
+    Instant gcTime = window.maxTimestamp().plus(windowingStrategy.getAllowedLateness());
+    Instant inputWM = stepContext.timerInternals().currentInputWatermarkTime();
+    if (gcTime.isBefore(inputWM)) {
+      // The element is too late for this window.
+      droppedDueToLateness.addValue(1L);
+      WindowTracing.debug(
+          "StatefulDoFnRunner.processElement/onTimer: Dropping element for window:{} "
+              + "since too far behind inputWatermark:{}; outputWatermark:{}",
+          window, timerInternals.currentInputWatermarkTime(),
+          timerInternals.currentOutputWatermarkTime());
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public void onTimer(
+      String timerId, BoundedWindow window, Instant timestamp, TimeDomain timeDomain) {
+    StateInternals<?> stateInternals = stepContext.stateInternals();
```

Comment:
Dual to the above idea about `CleanupTimer`, here you could have

```java
interface StateCleaner {
  void clearForWindow(BoundedWindow, DoFnSignature)
}
```

And provide the code here as a default implementation based on `StateInternals`.

Then this `StatefulDoFnRunner` class becomes very small and focused - it just owns the pattern of setting and responding to the timer but none of the details. And all the same support code is available to runner authors via the default implementations.